csrc/scheduler/reduction.cpp

```diff
@@ -507,29 +507,270 @@ std::shared_ptr<ReductionParams> innerReductionHeuristic(
+  const float min_sm_efficiency = 0.9f;
+  float f_wave = (float)hp.gidim / (float)sm_count;
+  float sm_efficiency = f_wave / std::ceil(f_wave);
+  if (sm_efficiency >= min_sm_efficiency || small_reduction) {
+    return std::make_optional(hp);
+  } else {
+    return std::nullopt;
+  }
+}
+
 std::shared_ptr<ReductionParams> outerReductionHeuristic(
     const int64_t total_reduction_numel,
     const int64_t total_iteration_numel,
     const int64_t n_tensor_inputs,
+    const int64_t n_tensor_outputs,
     const int64_t max_input_dtype_size,
     const size_t vectorize_factor) {
   // WARNING: Current device for codegen may not be the target device
+  auto dev_prop = at::cuda::getCurrentDeviceProperties();
   const int64_t device_max_threads_per_multiprocessor =
-      (int64_t)at::cuda::getCurrentDeviceProperties()
-          ->maxThreadsPerMultiProcessor;
+      (int64_t)dev_prop->maxThreadsPerMultiProcessor;
 
   const int64_t device_multiprocessor_count =
-      (int64_t)at::cuda::getCurrentDeviceProperties()->multiProcessorCount;
-
+      (int64_t)dev_prop->multiProcessorCount;
+
+  // Set register used to store vectorized and unrolled data loaded from gmem.
+  // A large value allows more unroll and vectorization, which is beneficial
+  // for memory-bound kernels. However, it increases register pressure and may
+  // lead to lower occupancy which is bad for compute-bound kernels. In most
+  // cases, the scheduler uses 512 threads and to reach an occupancy of 50%,
+  // each thread can use up to 64 registers, here only 8 registers are reserved
+  // for unroll and vectorization. The fused ops can have 48 registers for other
+  // purposes. Test shows it leads to 50% occupancy for outer reduction without
+  // fused ops and 50% occupancy for gelu backward which fused 21 ops including
+  // the expensive tanh op. Further tuning of this heuristic can utilize the
+  // cost of the fused ops.
+  const int64_t buffer_reg_count = 8L;
   auto const max_unroll = ceilDiv(
       // Available unrolling based on size of data type
-      (int64_t)16 / (int64_t)max_input_dtype_size,
+      buffer_reg_count * scheduler_utils::bytes_per_register /
+          (int64_t)max_input_dtype_size,
       // Reduce unrolling if we have many inputs, start reduction at 4 inputs
       scheduler_utils::lastPow2(
           std::max((int64_t)n_tensor_inputs >> 2, (int64_t)1)));
 
+  // Try block reduction
+  const int64_t max_threads_per_block = (int64_t)dev_prop->maxThreadsPerBlock;
+  std::optional<outerReduHeuristicParas> maybe_blk_hp =
+      maybeBlockOuterReduction(
+          (int64_t)vectorize_factor,
+          device_multiprocessor_count,
+          max_unroll,
+          max_threads_per_block,
+          n_tensor_outputs,
+          total_iteration_numel,
+          total_reduction_numel);
+  if (maybe_blk_hp.has_value()) {
+    return heuristicParaToSchedulerPara(maybe_blk_hp.value());
+  }
```

Comment:
Note for myself: If the condition returns false here, I think we'll just go along with the old logic on block reduction...